src/service/lightbox-manager-impl.js

```diff
@@ -0,0 +1,168 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {whenDocumentReady} from '../document-ready';
+import {isExperimentOn} from '../experiments';
+import {autoDiscoverLightboxables} from './lightbox-manager-discovery';
+import {dev} from '../log';
+import {fromClassForDoc} from '../service';
+import {timerFor} from '../timer';
+
+
+/**
+ * LightboxManager is a document-scoped service responsible for:
+ *  -Finding elements marked to be lightboxable (via `lightbox` attribute)
+ *  -Keeping an ordered list of lightboxable elements
+ *  -Providing functionality to get next/previous lightboxable element given
+ *   the current element.
+ *  -Discovering elements that can be auto-lightboxed and add the
+ *   `lightbox` attribute and possibly an on-tap handler to them
+ */
+class LightboxManager {
+
+  /**
+   * @param {!./ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+
+    // Extra safety check, we don't install this service if experiment is off
+    dev().assert(isExperimentOn(ampdoc.win, 'amp-lightbox-viewer'));
+
+    /** @const @private {!./ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /**
+     * Ordered list of lightboxable elements
+     * @private {!Array<!Element>}
+     **/
+    this.elements_ = null;
+
+    /**
+     * Cache for the `maybeInit()` call.
+     * @private {Promise}
+     **/
+    this.initPromise_ = null;
+
+    // TODO(aghassemi): Find a better time to initialize, maybe after the first
```

Comment:
Good question. We've been planning to add `Resources.runWhenIdleish()` method for a while...